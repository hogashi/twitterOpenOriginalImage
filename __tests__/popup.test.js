import TestRenderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
const { shallow } = Enzyme;

import {
  OPTIONS_TEXT,
  SHOW_ON_TIMELINE,
  isFalse,
  SHOW_ON_TWEETDECK_TIMELINE,
  isTrue,
  OPTION_KEYS,
} from '../src/main';
import { Popup } from '../src/components/Popup';

describe('Popup', () => {
  it('render', () => {
    const optionsText = OPTIONS_TEXT;
    const optionKeys = OPTION_KEYS;
    const options = {};
    optionKeys.forEach(key => {
      options[key] = isTrue;
    });

    const props = {
      optionsText,
      optionKeys,
      options,
    };
    const tree = TestRenderer.create(<Popup {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('保存ボタン押すと設定が保存される', () => {
    window.localStorage = {};
    it('最初は空', () => {
      expect(window.localStorage).toMatchObject({});
    });

    const optionsText = OPTIONS_TEXT;
    const optionKeys = OPTION_KEYS;
    const options = {};
    const expectOptions = {};
    optionKeys.forEach(key => {
      options[key] = isTrue;
      expectOptions[key] = isTrue;
    });
    // 初期設定いっこOFFにしてみる
    options[SHOW_ON_TIMELINE] = isFalse;
    expectOptions[SHOW_ON_TIMELINE] = isFalse;

    const props = {
      optionsText,
      optionKeys,
      options,
    };

    window.chrome = {
      tabs: {
        query: jest.fn((_, callback) => {
          callback([
            {
              // 対象タブ
              id: 1,
              url: 'http://twitter.com',
            },
            {
              // 対象ではないタブ
              id: 1,
              url: 'http://google.com',
            },
            {
              // 対象ではないタブ
              id: 1,
            },
            {
              // 対象ではないタブ
              url: 'http://twitter.com',
            },
          ]);
        }),
        sendMessage: jest.fn((id, option, callback) => {
          callback('mock ok');
        }),
      },
    };
    const wrapper = shallow(<Popup {...props} />);

    it('渡した設定がそのまま保存される', () => {
      wrapper.find('.saveSettingButton').simulate('click');
      // 送りたいタブは正しい形式かつ対象ホストなタブのみ
      expect(window.chrome.tabs.query.mock.calls.length).toBe(1);

      expect(window.localStorage).toMatchObject(expectOptions);
    });

    it('チェックボックスをクリックして保存すると設定変えられる', () => {
      wrapper.find(`.${SHOW_ON_TIMELINE}`).simulate('click');
      wrapper.find(`.${SHOW_ON_TWEETDECK_TIMELINE}`).simulate('click');
      expectOptions[SHOW_ON_TIMELINE] = isTrue;
      expectOptions[SHOW_ON_TWEETDECK_TIMELINE] = isFalse;

      wrapper.find('.saveSettingButton').simulate('click');
      expect(window.localStorage).toMatchObject(expectOptions);
    });

    it('何度も設定変えられる', () => {
      wrapper.find(`.${SHOW_ON_TIMELINE}`).simulate('click');
      expectOptions[SHOW_ON_TIMELINE] = isFalse;

      wrapper.find('.saveSettingButton').simulate('click');
      expect(window.localStorage).toMatchObject(expectOptions);
    });
  });
});
