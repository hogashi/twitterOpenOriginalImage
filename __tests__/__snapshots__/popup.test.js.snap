// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup render 1`] = `
<div
  style={
    {
      "display": "flex",
      "flexFlow": "column",
      "justifyContent": "center",
      "minWidth": "300px",
    }
  }
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
      style={
        {
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      <h5
        className="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
        style={
          {
            "flex": "0 1 auto",
          }
        }
      >
        Options - 設定
      </h5>
    </div>
  </header>
  <ul
    className="MuiList-root MuiList-padding MuiList-subheader"
  >
    <div
      className="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
    >
      TwitterWeb公式
    </div>
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiListItem-root checkboxListItem SHOW_ON_TIMELINE MuiListItem-dense MuiListItem-gutters MuiListItem-button"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "padding": "4px 12px",
          }
        }
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={true}
            className="PrivateSwitchBase-input-4"
            data-indeterminate={false}
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
        </span>
      </span>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          タイムラインにボタンを表示
        </span>
      </div>
    </div>
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiListItem-root checkboxListItem SHOW_ON_TWEET_DETAIL MuiListItem-dense MuiListItem-gutters MuiListItem-button"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "padding": "4px 12px",
          }
        }
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={true}
            className="PrivateSwitchBase-input-4"
            data-indeterminate={false}
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
        </span>
      </span>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          ツイート詳細にボタンを表示
        </span>
      </div>
    </div>
  </ul>
  <ul
    className="MuiList-root MuiList-padding MuiList-subheader"
  >
    <div
      className="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
    >
      TweetDeck
    </div>
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiListItem-root checkboxListItem SHOW_ON_TWEETDECK_TIMELINE MuiListItem-dense MuiListItem-gutters MuiListItem-button"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "padding": "4px 12px",
          }
        }
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={true}
            className="PrivateSwitchBase-input-4"
            data-indeterminate={false}
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
        </span>
      </span>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          タイムラインにボタンを表示
        </span>
      </div>
    </div>
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiListItem-root checkboxListItem SHOW_ON_TWEETDECK_TWEET_DETAIL MuiListItem-dense MuiListItem-gutters MuiListItem-button"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "padding": "4px 12px",
          }
        }
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={true}
            className="PrivateSwitchBase-input-4"
            data-indeterminate={false}
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
        </span>
      </span>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          ツイート詳細にボタンを表示
        </span>
      </div>
    </div>
  </ul>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained saveSettingButton MuiButton-containedPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      {
        "flex": "0 1 auto",
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      設定を保存
    </span>
  </button>
  <footer
    style={
      {
        "marginTop": "3px",
        "textAlign": "center",
      }
    }
  >
    twitter画像原寸ボタン - hogashi
  </footer>
</div>
`;
